lib/semantics/check-do-stmt.cc

```diff
@@ -406,31 +410,50 @@ class DoStmtContext {
       // TODO: Mark the following message as a warning when we have warnings
       context_.Say(sourceLocation, "DO controls should be INTEGER"_en_US);
     } else {
-      context_.Say(sourceLocation, "DO controls should be INTEGER"_err_en_US);
+      SayBadDoControl(sourceLocation);
     }
   }
 
   void CheckDoVariable(const parser::ScalarName &scalarName) {
-    const DeclTypeSpec *symType{scalarName.thing.symbol->GetType()};
-    if (symType->IsNumeric(TypeCategory::Integer)) {
-      return;  // No warnings or errors for INTEGER
-    }
     const parser::CharBlock &sourceLocation{scalarName.thing.source};
-    CheckDoControl(sourceLocation, symType->IsNumeric(TypeCategory::Real));
+    const Symbol *symbol{scalarName.thing.symbol};
+    if (!symbol) {
+      SayBadDoControl(sourceLocation);
+    } else {
+      if (!IsVariableName(*symbol)) {
+        context_.Say(
+            sourceLocation, "DO control must be an INTEGER variable"_err_en_US);
+      } else {
+        const DeclTypeSpec *symType{symbol->GetType()};
+        if (!symType) {
+          SayBadDoControl(sourceLocation);
+        } else {
+          if (!symType->IsNumeric(TypeCategory::Integer)) {
+            CheckDoControl(
+                sourceLocation, symType->IsNumeric(TypeCategory::Real));
+          }
+        }
+      } // No messages for INTEGER
+    }
   }
 
   void CheckDoExpression(const parser::ScalarExpr &scalarExpression) {
     const evaluate::Expr<evaluate::SomeType> *expr{GetExpr(scalarExpression)};
-    if (ExprHasTypeCategory(*expr, TypeCategory::Integer)) {
-      return;  // No warnings or errors for INTEGER
-    }
     const parser::CharBlock &sourceLocation{
         scalarExpression.thing.value().source};
-    CheckDoControl(
-        sourceLocation, ExprHasTypeCategory(*expr, TypeCategory::Real));
+    if (!expr) {
+      SayBadDoControl(sourceLocation);
```

Comment:
Same thing here: expression analysis should have reported an error.

Reply:
There are tests for this in dosemantics03.f90.  This situation happens when you have an invalid value for an expression.  In this case, other error messages also appear.  So I really didn't need to emit an additional message, but I **did** need to check for a null value for expr.  Otherwise, the
compiler seg faults.

I added tests for the initial, final, and step expressions at lines 222, 261, and 300.

Both of these cases were pointed out by Jean.